### PR TITLE
overlord: allow seeding in the case of classic with modes system

### PR DESCRIPTION
**Pull request by alfonsosanchezbeato**

### CODE REVIEW

File: overlord/devicestate/devicemgr.go [hunk 1]
```diff
@@ -144,7 +147,10 @@ func Manager(s *state.State, hookManager *hookstate.HookManager, runner *state.T
 			return nil, err
 		}
 		if modeEnv != nil {
+			logger.Debugf("modeenv for model %s found", modeEnv.Model)
```
Comment (mardy) [inline overlord/devicestate/devicemgr.go:150]: nitpick: just noticed that on one comment below you spelt this as "modeEnv".

**Review by mardy — APPROVED** [review]
LGTM, thanks!

File: seed/seed16.go [hunk 2]
```diff
@@ -61,6 +62,7 @@ type seed16 struct {
 }
 
 func (s *seed16) LoadAssertions(db asserts.RODatabase, commitTo func(*asserts.Batch) error) error {
+	logger.Debugf("loading assertions for UC16/18 system")
```
Comment (pedronis) [inline seed/seed16.go:65]: this is clear enough from the logging loadAssertions, no?

File: seed/seed20.go [hunk 3]
```diff
@@ -82,6 +83,7 @@ type seed20 struct {
 }
 
 func (s *seed20) LoadAssertions(db asserts.RODatabase, commitTo func(*asserts.Batch) error) error {
+	logger.Debugf("loading assertions for core boot system")
```
Comment (pedronis) [inline seed/seed20.go:86]: same

File: overlord/devicestate/devicestate_test.go [hunk 4]
```diff
@@ -847,6 +884,40 @@ func (s *deviceMgrSuite) TestCheckKernel(c *C) {
 	c.Check(err, ErrorMatches, `cannot install "krnl_foo", parallel installation of kernel or gadget snaps is not supported`)
 }
 
+func (s *deviceMgrSuite) TestCheckKernelOnClassic(c *C) {
```
Comment (pedronis) [inline overlord/devicestate/devicestate_test.go:887]: ClassicWithModes ?

File: overlord/devicestate/firstboot.go [hunk 5]
```diff
@@ -327,7 +329,7 @@ func populateStateFromSeedImpl(st *state.State, opts *populateStateFromSeedOptio
 	return tsAll, nil
 }
 
-func importAssertionsFromSeed(st *state.State, sysLabel string) (seed.Seed, error) {
+func importAssertionsFromSeed(st *state.State, sysLabel string, hasModeEnv bool) (seed.Seed, error) {
```
Comment (pedronis) [inline overlord/devicestate/firstboot.go:332]: my preference would be to make the extra bool we pass to be isCoreBoot

File: overlord/devicestate/firstboot.go [hunk 6]
```diff
@@ -337,8 +339,8 @@ func importAssertionsFromSeed(st *state.State, sysLabel string) (seed.Seed, erro
 	// collect and
 	// set device,model from the model assertion
 	deviceSeed, err := loadDeviceSeed(st, sysLabel)
-	if err == seed.ErrNoAssertions && release.OnClassic {
-		// on classic seeding is optional
+	if err == seed.ErrNoAssertions && !hasModeEnv && release.OnClassic {
```
Comment (pedronis) [inline overlord/devicestate/firstboot.go:342]: then here we can use !isCoreBoot

File: overlord/devicestate/firstboot.go [hunk 7]
```diff
@@ -337,8 +339,8 @@ func importAssertionsFromSeed(st *state.State, sysLabel string) (seed.Seed, erro
...
+		// if classic boot seeding is optional
 		// set the fallback model
 		err := setClassicFallbackModel(st, device)
 		if err != nil {
```
Comment (pedronis) [inline overlord/devicestate/firstboot.go:346]: there is code later in this function that does:

release.OnClassic != classicModel

that will not be correct in modes != run for a classic with modes system, maybe not to fix here yet but probably worth adding at least a comment there

File: overlord/devicestate/devicemgr.go (hunk 1 above)
Comment (pedronis) [inline overlord/devicestate/devicemgr.go:150]: s/%s/%q/ to be on the safe side

File: overlord/devicestate/devicestate_test.go [hunk 8]
```diff
@@ -469,6 +469,38 @@ func (s *deviceMgrSuite) TestDeviceManagerEnsureBootOkSkippedOnNonRunModes(c *C)
 	c.Assert(err, IsNil)
 }
 
+func (s *deviceMgrSuite) switchDevManagerToClassicWithModes(c *C) {
+	release.OnClassic = true
+
+	// mock the modeenv file
+	m := boot.Modeenv{
+		Mode:           "run",
+		RecoverySystem: "20191127",
```
Comment (pedronis) [inline overlord/devicestate/devicestate_test.go:478]: this is only set on first boot, so not sure we always want it set for realism. One option is pass it in as needed.

**Review by pedronis — COMMENTED** [review]
thanks, did a pass

File: seed/seed16.go (hunk 2 above)
Comment (alfonsosanchezbeato) [reply seed/seed16.go:65]: In general I am trying to add some more debug traces here and there as I think we are not very verbose and sometimes it gets difficult to get good logs from customers. In the end you will not see this in a normal run. But yeah, maybe these two ones are not that useful.

File: overlord/devicestate/firstboot.go (hunk 5 above)
Comment (alfonsosanchezbeato) [reply overlord/devicestate/firstboot.go:332]: Changed

File: overlord/devicestate/firstboot.go (hunk 7 above)
Comment (alfonsosanchezbeato) [reply overlord/devicestate/firstboot.go:346]: Added comment

File: overlord/devicestate/devicemgr.go (hunk 1 above)
Comment (alfonsosanchezbeato) [reply overlord/devicestate/devicemgr.go:150]: Right, in general it looks like modeenv is preferred, so I'll change things in that direction.

Comment (alfonsosanchezbeato) [reply overlord/devicestate/devicemgr.go]: Done

File: overlord/devicestate/devicestate_test.go (hunk 8 above)
Comment (alfonsosanchezbeato) [reply overlord/devicestate/devicestate_test.go]: Removed

File: overlord/devicestate/devicestate_test.go (hunk 4 above)
Comment (alfonsosanchezbeato) [reply overlord/devicestate/devicestate_test.go]: Changed

**Review by alfonsosanchezbeato — COMMENTED** [review]
Thanks for the reviews, I've addressed the comments now.

File: overlord/devicestate/firstboot.go [hunk 9]
```diff
@@ -97,11 +97,11 @@ func populateStateFromSeedImpl(st *state.State, opts *populateStateFromSeedOptio
 	mode := "run"
 	sysLabel := ""
 	preseed := false
-	hasModeenv := false
+	isCoreBoot := false
 	if opts != nil {
 		if opts.Mode != "" {
 			mode = opts.Mode
-			hasModeenv = true
+			isCoreBoot = true
```
Comment (pedronis) [inline overlord/devicestate/firstboot.go:104]: maybe it was confusing, but my ask was not to just change the name, this is not the definition of core boot, we also need to set if !release.OnClassic

File: overlord/devicestate/firstboot.go [hunk 10]
```diff
@@ -97,11 +97,11 @@ func populateStateFromSeedImpl(st *state.State, opts *populateStateFromSeedOptio
 	mode := "run"
 	sysLabel := ""
 	preseed := false
-	hasModeenv := false
+	isCoreBoot := false
```
Comment (pedronis) [inline overlord/devicestate/firstboot.go:100]: this should be:

isCoreBoot = !release.OnClassic

then the setting to true below is also correct

File: overlord/devicestate/firstboot.go [hunk 11]
```diff
@@ -339,7 +339,7 @@ func importAssertionsFromSeed(st *state.State, sysLabel string, hasModeEnv bool)
 	// collect and
 	// set device,model from the model assertion
 	deviceSeed, err := loadDeviceSeed(st, sysLabel)
-	if err == seed.ErrNoAssertions && !hasModeEnv && release.OnClassic {
+	if err == seed.ErrNoAssertions && !isCoreBoot && release.OnClassic {
```
Comment (pedronis) [inline overlord/devicestate/firstboot.go:342]: with the correctly defined isCoreBoot we don't need the `&& release.OnClassic` part anymore

**Review by pedronis — COMMENTED** [review]
some tweaks are still necessary

File: overlord/devicestate/firstboot.go (hunk 7 above)
Comment (pedronis) [reply overlord/devicestate/firstboot.go:346]: thanks

File: overlord/devicestate/firstboot.go (hunk 9 above)
Comment (alfonsosanchezbeato) [reply overlord/devicestate/firstboot.go]: Oh, sorry, I just renamed without stopping one sec for thinking. Changed.

File: overlord/devicestate/firstboot.go (hunk 10 above)
Comment (alfonsosanchezbeato) [reply overlord/devicestate/firstboot.go]: I've tweaked this in a slightly different way that I think is more readable, let me know if you prefer what you suggest.

**Review by alfonsosanchezbeato — COMMENTED** [review]
@pedronis thanks, changed things in the suggested direction

File: overlord/devicestate/devicemgr.go (hunk 1 above)
Comment (mardy) [reply overlord/devicestate/devicemgr.go]: But you didn't change it yet, did you?

Comment (alfonsosanchezbeato) [reply overlord/devicestate/devicemgr.go]: I did for this concrete variable! But not for the full PR, I've done that now :)

**Review by mardy — APPROVED** [review]
Thanks for the updates.

File: overlord/devicestate/firstboot.go (hunk 10 above)
Comment (pedronis) [reply overlord/devicestate/firstboot.go]: it's ok, my thinking is that we might need isCoreBoot in other parts of the function later, but it's a hypothetical at this point, so it's fine either way, we can changed again later, thanks

**Review by pedronis — APPROVED** [review]
thank you

**Pull request merged**